### PR TITLE
Ca/add and or

### PR DESCRIPTION
**Pull request by castengo**

### CODE REVIEW

File: src/string_instruction_parser.yrl [hunk 1]
```diff
@@ -1,43 +1,39 @@
...
-predicate -> lit comparator array : [unwrap('$1'), [<<"array">>, '$3'], unwrap('$2')].
-predicate -> string comparator array : [unwrap_string('$1'), [<<"array">>, '$3'], unwrap('$2')].
-predicate -> string comparator string : [unwrap_string('$1'), unwrap_string('$3'), unwrap('$2')].
-predicate -> load comparator string : [unwrap('$1'), unwrap_string('$3'), unwrap('$2')].
+predicates -> predicate_group : '$1'.
+predicates -> bang predicate_group : lists:append('$2', [[<<"not">>]]).
+
+predicate_group -> predicate : '$1'.
+predicate_group -> predicate 'and' predicates : lists:append('$1', [jump(jfalse, '$3') | '$3']).
+predicate_group -> predicate 'or' predicates : lists:append('$1', [jump(jtrue, '$3') | '$3']).
```
Comment (castengo) [inline src/string_instruction_parser.yrl:15]: I feel like there is a prettier way to do this but my brain hurts and this works

File: src/string_instruction_parser.yrl [hunk 2]
```diff
@@ -1,43 +1,39 @@
...
+
+predicate -> lit : [unwrap('$1')].
+predicate -> load : [unwrap('$1'), [<<"to_bool">>]].
+predicate -> variable endcomparator : ['$1', unwrap('$2')].
+predicate -> variable comparator variable : ['$1', '$3', unwrap('$2')].
+predicate -> variable between lit 'and' lit : ['$1', unwrap('$3'), unwrap('$5'), unwrap('$2')].
+predicate -> variable comparator array : ['$1', [<<"array">>, '$3'], unwrap('$2')].
+
+variable -> lit : unwrap('$1').
+variable -> load : unwrap('$1').
```
Comment (castengo) [inline src/string_instruction_parser.yrl:25]: cleaned this up a little, if I didn't I would have had to add `bang` to every possibility and the amount of matches doubled 😱

File: test/predicator_test.exs [hunk 3]
```diff
@@ -5,182 +5,439 @@ defmodule PredicatorTest do
...
 
-      assert {:ok, [["load", "age"], ["lit", 5], ["lit", 10], ["comparator", "BETWEEN"]]} =
-               Predicator.compile("age between 5 and 10")
+      assert {:ok, [[:load, :age], [:lit, 5], [:lit, 10], [:comparator, :BETWEEN], [:not]]} =
+               Predicator.compile("!age between 5 and 10", :atom_key_inst)
+    end
+
+    test "compiles string key instructions" do
+      assert {:ok, [["load", "age"], ["lit", 5], ["lit", 10], ["comparator", "BETWEEN"], ["not"]]} =
+               Predicator.compile("!age between 5 and 10")
```
Comment (castengo) [inline test/predicator_test.exs:54]: Added the `!` possibility to every type of predicate 😎 more tests the better, right? 😅 

Also, more tests was the only sane way to refactor the lex and parse files.

File: src/string_instruction_parser.yrl (hunk 1 above)
Comment (castengo) [reply src/string_instruction_parser.yrl:15]: I tried `['$1' | blah blah]` but it adds an extra pair of damn brackets

**Review by jrichocean — APPROVED** [review]

**Pull request merged**